### PR TITLE
Add helper to track usage of types from operation, interface, namespace

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix https://github.com/Azure/cadl-azure/issues/1733

### CODE REVIEW

File: packages/compiler/test/helpers/usage-resolver.test.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  });
+
+  it("track model referenced in union in returnType", async () => {
+    const usages = await getUsages(`
+      model Bar {}
+      model Foo {}
+      op test(): Foo | Bar;
+    `);
+
+    deepStrictEqual(usages, { inputs: [], outputs: ["Foo", "Bar"] });
```
Comment (markcowl) [inline packages/compiler/test/helpers/usage-resolver.test.ts:116]: Do we need to include a test of models that extend the types in the union variant?

File: packages/compiler/core/helpers/usage-resolver.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  isUsedAs(type: TrackableType, usage: UsageFlags): boolean;
+}
+
+/**
+ * Resolve usage(input, output or both) of various types in the given namespace.
+ * Will recursively scan all namespace, interfaces and operations contained inside the namespace.
+ * @param namespace Entrypoint namespace to get usage from.
+ * @returns Map of types to usage.
+ */
+export function resolveUsages(namespace: NamespaceType): UsageTracker {
```
Comment (markcowl) [inline packages/compiler/core/helpers/usage-resolver.ts:32]: Based on the issue, it feels like we need to provide a specific exported function that will return a list of models that are used in a specific operation (the input/output flags are nice), in addition to this function which answers the question over all operations in the namespace.

**Review by markcowl — CHANGES_REQUESTED** [review]
I think they may need this information more granularly at the operation level, although I think the namespace level is more useful in most generationcases.

File: packages/compiler/test/helpers/usage-resolver.test.ts (hunk 1 above)
Comment (timotheeguerin) [reply packages/compiler/test/helpers/usage-resolver.test.ts:116]: not sure what you mean by this test

File: packages/compiler/test/helpers/usage-resolver.test.ts [hunk 3]
```diff
@@ -124,4 +129,40 @@ describe("compiler: helpers: usage resolver", () => {
 
     deepStrictEqual(usages, { inputs: [], outputs: ["MyEnum"] });
   });
+
+  describe("resolving usage of specific operation", () => {
+    it("only find usage in that operation", async () => {
+      const usages = await getUsages(
+        `
+          model Foo {}
```
Comment (markcowl) [inline packages/compiler/test/helpers/usage-resolver.test.ts:137]: should include one model outside the op/interface in these tests.  Otherwise, lgtm

**Review by markcowl — APPROVED** [review]

**Pull request merged**